### ISSUE TITLE
Provide an API to show number of allocations per thread

### ISSUE THREAD

**Issue by karmeli87** [dotnet/runtime#40585]
## Background and Motivation

We develop a distributed application that on one hand process many items in batches and on the other send pings to its peers.
Processing of a single batch can generate a lot of objects, keep them around and eventually persist them on the disk.

Keeping around millions of objects will cause us GC pauses of ~500ms, which will destabilize the distributed application (it expect to receive ping every 300ms or so)

Since we use 3rd party lib during the process, we can't really track ourselves the number of the generated objects and limit the batch accordingly.

So having the ability to _estimate_ how many objects were allocated can help us break the batch execution, persist it and start new batch, without putting much pressure on the GC.

There is already something similar in form of `GC.GetAllocatedBytesForCurrentThread()`, but this will give the total size and not the number of objects. 

## Proposed API

I'm not fully aware of the impl. difficulty of something like this, but have the following API would be nice

`GC.GetAllocatedObjectsForCurrentThread()`

As a complementary to the `GC.GetAllocatedBytesForCurrentThread()`

## Usage Examples

```csharp
var initial = GC.GetAllocatedObjectsForCurrentThread();
while(true)
{ 
  // do work
  var current = GC.GetAllocatedObjectsForCurrentThread();
  if (current - inital > 1M)
    break;
}
````

**Comment by svick** [comment]
Why do you want to know the number of allocated objects instead of allocated bytes, when GC performance is mostly based on bytes?

**Comment by karmeli87** [comment]
As far as I understand the GC performance (at least the marking phase) is directly correlates to the number of objects rather to their size.

**Comment by ayende** [comment]
The GC performance is not actually based on bytes. Most of the GC costs is when you scan for objects, which is directly related to the number of objects it has to run.
For example, allocating 10 `byte[1024*1024]` is a lot cheaper than 1,000,000 `new byte[10]`.

**Comment by omariom** [comment]
> the GC performance (at least the marking phase) is directly correlates to the number of objects

The total number of reachable objects. Not the recently allocated ones.

**Comment by ayende** [comment]
@omariom In our scenario, we don't control how many reachable objects there are.
But many of them will be reachable. I don't want to have GC run when I have millions of reachable objects. I want to cut the batch short, which will us discard a lot of those objects, then a GC won't run into long pauses.

**Comment by jkotas** [comment]
The existing `GC.GetAllocatedBytesForCurrentThread` API was free to provide. The GC allocator was keeping track of amount allocated per thread already.

The GC allocator does not keep track of number of objects allocated today. If we were to provide this API, it would slow down the allocation path for everybody. It makes it tough tradeoff, especially given that the number of objects allocated is not very useful number as others have pointed out. 

> > the GC performance (at least the marking phase) is directly correlates to the number of objects

> The total number of reachable objects. Not the recently allocated ones.

And the total number of references that the GC needs to visit. For example, `object[100]` is a lot more expensive for marking phase than `IntPtr[100]`.

**Comment by ayende** [comment]
There is the `alloc_count` on the `gc_alloc_context`, isn't that keeping track of the count?

**Comment by jkotas** [comment]
`alloc_count` does not track individual allocations. `alloc_count` tracks how many times `allocate_more_space` got called for the purpose of heap balancing.

**Comment by kkokosa** [comment]
So, what you really want is to have a deterministic (or at least "capped") GC pauses. This sounds more like "make the GC more aggressive" story, while API you suggest seems to be a pretty detailed, specific solution. With not so obvious consequences, as stated before - as, AFAIK, the GC pause relation to any combination of number of allocated object, number of reachable objects or number of inter-objects references is not so obvious to make a good decision based on them.

So have you considered "make GC more aggresive" approach? By tuning `GCHeapCount` or hard limits? IMHO it would be interesting to understand your setup - GC mode, gen sizes etc.

**Comment by ayende** [comment]
My issue is that I have a bunch of process that allocate a lot of memory (indexing).
During the indexing run, the memory is referenced, once the indexing batch is completed, most of the memory can go away.
The problem is that I have no good way to tell _when_ I should close the batch. 

Closing the batch means that the GC doesn't have to scan so many objects, so I don't actually want the GC more aggressive, because that would simply make the problem worse. During the batch, I'm holding to the memory, so more GC runs would simply mean scanning more objects, promoting them to gen1/2, etc.

My primary issue is that the indexing function is provided by the user, so I don't know what allocations they are going to be doing there.

**Comment by jkotas** [comment]
> The problem is that I have no good way to tell when I should close the batch.

What is the metrics you want to optimize here in priority order? Peak memory consumption? % time spent in GC? Maximum GC pause times?

**Comment by ayende** [comment]
Maximum GC pause, yes.

The problem actually happened on a machine with 128GB of RAM, and the system was using maybe 6GB ?

**Comment by ayende** [comment]
Another possibility that come to mind is that we have a GC pause because we have had gen0 collections that couldn't collect the values, since they were used during indexing.
That move them to Gen1/2, and make the collection harder.

**Comment by jkotas** [comment]
I sounds like you want to prepare the system for the GC before it actually kicks in. We have APIs like that already: https://docs.microsoft.com/en-us/dotnet/api/system.gc.waitforfullgcapproach?view=netcore-3.1#System_GC_WaitForFullGCApproach . I am not sure whether the existing APIs will work for your specific case, but APIs like this would be solution to the problem you are describing (assuming the solution is a new API - it is not clear as @kkokosa pointed out).

**Comment by ayende** [comment]
I don't think that this is a correct way to express what I want. 
I don't actually care when the GC is going to happen, only that it isn't going to cause a pause.
If I were to be notified that it is approaching, I would be able to stop the batch, but stopping the batch still means that I have to do some (non trivial) amount of processing. 

That in turn means that it is likely that by the time the GC runs, I'll still be busy with the batch's data.

Basically, I want to have a good way to control my batch size so none of them would overwhlme the GC
